### PR TITLE
FleshWard and Mists

### PR DESCRIPTION
**Pull request by petercombs**

### CODE REVIEW

File: eldritch/test_usables.py [hunk 1]
```diff
@@ -374,6 +376,101 @@ def testCannotBeCastOutsideOtherworldMovement(self):
   # TODO: a test covering the ability to cast after travelling during the movement phase
 
 
+class FleshWardTest(EventTest):
+  def setUp(self):
```
Comment (fingoldfish) [inline eldritch/test_usables.py:380]: Two additional tests that I would recommend - one where you use a Food to reduce the stamina loss from 1 to 0, and then validate that flesh ward is not castable.
The other one would be a test where you fail to cast flesh ward, and validate that it is exhausted and no longer usable.

File: eldritch/test_usables.py [hunk 2]
```diff
@@ -419,5 +516,43 @@ def testMultipleOptions(self):
...
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=4)):
+      self.resolve_to_usable(0, "Mists0", events.CastSpell)
+    self.state.event_stack.append(self.state.usables[0]["Mists0"])
+    choice = self.resolve_to_choice(MultipleChoice)
+    choice.resolve(self.state, "Cast")
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=5)):
+      spend_choice = self.resolve_to_choice(SpendChoice)
+    spend_choice.resolve(self.state, "Done")
+    self.resolve_until_done()
+
```
Comment (fingoldfish) [inline eldritch/test_usables.py:553]: Same comment as above - recommend at least one test where they fail to cast, and you validate that it's exhausted and no longer usable.

**Review by fingoldfish — CHANGES_REQUESTED** [review]

File: eldritch/test_usables.py (hunk 1 above)
Comment (fingoldfish) [reply eldritch/test_usables.py:380]: Addendum: add a TODO for a test to discard this if the ancient one awakens.

File: eldritch/items.py [hunk 3]
```diff
@@ -347,6 +347,69 @@ def deactivate(self):
...
+class Mists(Spell):
+  def __init__(self, idx):
+    super().__init__("Mists", idx, {}, 0, None, 0)
+    self.evade = None
+
+  def get_usable_interrupt(self, event, owner, state):
+    if event.is_done() or self.exhausted or getattr(event, "character", None) != owner:
+      return None
+    # TODO: be able to cast Mists on an EvadeCheck
+    if event.is_done():
```
Comment (fingoldfish) [inline eldritch/items.py:391]: You check for event.is_done() twice.

File: eldritch/test_usables.py [hunk 4]
```diff
@@ -374,6 +376,148 @@ def testCannotBeCastOutsideOtherworldMovement(self):
...
+    self.assertEqual(self.char.stamina, 5)
+    self.assertEqual(self.char.sanity, 4)
+
+    fight_or_flee.resolve(self.state, "Fight")
+    combat_choice = self.resolve_to_choice(CombatChoice)
+    combat_choice.resolve(self.state, "done")
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=4)):
+      self.resolve_to_choice(MultipleChoice)
+      # no longer usable
+      self.assertNotIn("Flesh Ward0", self.state.usables)
```
Comment (fingoldfish) [inline eldritch/test_usables.py:413]: Nit: self.state.usables is a dict that maps idx -> a dict from handle -> event. So actually, you want to check that `0` is not in self.state.usables here. Or just asserting that it's empty is enough (as you do on the line below).

File: eldritch/test_usables.py [hunk 5]
```diff
@@ -374,6 +376,148 @@ def testCannotBeCastOutsideOtherworldMovement(self):
...
+    self.assertEqual(self.char.stamina, 4)
+    self.assertEqual(self.char.sanity, 4)
+
+    fight_or_flee.resolve(self.state, "Fight")
+    combat_choice = self.resolve_to_choice(CombatChoice)
+    combat_choice.resolve(self.state, "done")
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=4)):
+      self.resolve_to_choice(MultipleChoice)
+      # no longer usable
+      self.assertNotIn("Flesh Ward0", self.state.usables)
```
Comment (fingoldfish) [inline eldritch/test_usables.py:445]: See above.

File: eldritch/test_usables.py [hunk 6]
```diff
@@ -374,6 +376,148 @@ def testCannotBeCastOutsideOtherworldMovement(self):
...
+
+  def testNotUsableAfterDamageReduced(self):
+    self.char.possessions.append(items.Food(0))
+    self.assertEqual(self.char.stamina, 5)
+    self.assertEqual(self.char.sanity, 5)
+    self.state.event_stack.append(Loss(self.char, {"stamina": 1}))
+    self.resolve_to_usable(0, "Flesh Ward0", events.CastSpell)
+    self.state.event_stack.append(self.state.usables[0]["Food0"])
+    self.resolve_until_done()
+    self.assertNotIn("Flesh Ward0", self.state.usables)
```
Comment (fingoldfish) [inline eldritch/test_usables.py:515]: See above.

File: eldritch/test_usables.py [hunk 7]
```diff
@@ -419,5 +563,58 @@ def testMultipleOptions(self):
...
+    self.state.event_stack.append(EvadeOrCombat(self.char, monster))
+    evade_choice = self.resolve_to_choice(MultipleChoice)
+    evade_choice.resolve(self.state, "Evade")
+    self.resolve_to_usable(0, "Mists0", events.CastSpell)
+    self.state.event_stack.append(self.state.usables[0]["Mists0"])
+    choice = self.resolve_to_choice(MultipleChoice)
+    choice.resolve(self.state, "Cast")
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=4)):
+      self.resolve_to_choice(MultipleChoice)
+      self.assertNotIn("Mists0", self.state.usables)
```
Comment (fingoldfish) [inline eldritch/test_usables.py:595]: See above.

File: eldritch/test_usables.py [hunk 8]
```diff
@@ -419,5 +563,58 @@ def testMultipleOptions(self):
...
+    evade_choice.resolve(self.state, "Evade")
+    self.resolve_to_usable(0, "Mists0", events.CastSpell)
+    self.state.done_using[0] = True
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=4)):
+      self.resolve_to_usable(0, "Mists0", events.CastSpell)
+    self.state.event_stack.append(self.state.usables[0]["Mists0"])
+    choice = self.resolve_to_choice(MultipleChoice)
+    choice.resolve(self.state, "Cast")
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=5)):
+      spend_choice = self.resolve_to_choice(SpendChoice)
```
Comment (fingoldfish) [inline eldritch/test_usables.py:612]: After you successfully cast this, shouldn't the spend choice get cancelled?

**Review by fingoldfish — APPROVED** [review]

File: eldritch/test_usables.py (hunk 8 above)
Comment (petercombs) [reply eldritch/test_usables.py:611]: This new spend choice is to spend clues for the Mists spell check. Maybe you wanted to be extra misty.

**Pull request merged**